backend/src/apiserver/main.go

```diff
@@ -87,10 +91,10 @@ func startRpcServer(resourceManager *resource.ResourceManager) {
 		glog.Fatalf("Failed to start RPC server: %v", err)
 	}
 	s := grpc.NewServer(grpc.UnaryInterceptor(apiServerInterceptor), grpc.MaxRecvMsgSize(math.MaxInt32))
-	api.RegisterPipelineServiceServer(s, server.NewPipelineServer(resourceManager))
-	api.RegisterExperimentServiceServer(s, server.NewExperimentServer(resourceManager))
-	api.RegisterRunServiceServer(s, server.NewRunServer(resourceManager))
-	api.RegisterJobServiceServer(s, server.NewJobServer(resourceManager))
+	api.RegisterPipelineServiceServer(s, server.NewPipelineServer(resourceManager, &server.PipelineServerOptions{CollectMetrics: *collectMetricsFlag}))
+	api.RegisterExperimentServiceServer(s, server.NewExperimentServer(resourceManager, &server.ExperimentServerOptions{CollectMetrics: *collectMetricsFlag}))
+	api.RegisterRunServiceServer(s, server.NewRunServer(resourceManager, &server.RunServerOptions{CollectMetrics: *collectMetricsFlag}))
+	api.RegisterJobServiceServer(s, server.NewJobServer(resourceManager, &server.JobServerOptions{CollectMetrics: *collectMetricsFlag}))
 	api.RegisterReportServiceServer(s, server.NewReportServer(resourceManager))
```

Comment:
This is a very interesting part. When I tried to find the metrics for report server, e.g., the workflow gc (i.e., workflow client's Delete method), it turns out the code is in resource_manager. In other words, PA calls report server's api method, and report server's method calls resource manager's method. And to make things a little bit confusing is that, the method in report server is just a wrapper and the logic is in resource manager,  and thus if you want to measure a specific operatiion, you have to go into the resource manager code. E.g., report server has ReportScheduledWorkflow method. Would it make sense to count this request? Not really, since this method can either do the indeed reporting to the db or removing the workflow(and this logic is actually in resource manager....)

Reply:
I think it's useful counting how many failed